### DISCUSSION TITLE
Fail to bundle the gems (on rugged) while installing a dev env

### DISCUSSION THREAD

**Discussion by jpprost** [consuldemocracy/consuldemocracy#5568] [Q&A] [answered]
Hi,

I am going through the installation of a local development environment, and I am stuck in the middle of Step # 5 (install the required gems using Bundler), during the installation of rugged.
The output on stdout is in the file ```stdout.bundle.dev-env.txt``` attached.
The more detailed log file ```mkmf.log``` is also attached.

My system obviously misses something ("You have to install development tools first.").
While trying to interpret the output messages, I checked that the following packages were installed:
- ```libgssapi-krb5-2, libgssapi3-heimdal```
- various ```libpcreXXX```:
  - ```libpcre2-16-0```
  - ```libpcre2-32-0```
  - ```libpcre2-8-0```
  - ```libpcre3```
- ```libssh2-1, libssh2-1-dev```
- ```libgit2-28, libgit2-dev, libgit2-glib-1.0-dev```

I may miss some of the binding packages to one or more of thoses libraries, but since I do not know the underlying programming language(s) in use it is hard to figure out by myself which ones.

I have installed ```homebrew, gcc, rbenv```, and I have also double-checked the prerequisites for Ubuntu from https://docs.consuldemocracy.org/tech_docs/introduction-1/local_installation/ubuntu.

My local environment :
- PC Intel Core i7
- 62.5 GiB RAM
- Ubuntu 20.04

Any hints to put me back on track? That would be much appreciated :-)

Cheers,
JP
[stdout.bundle.dev-env.txt](https://github.com/user-attachments/files/15566333/stdout.bundle.dev-env.txt)
[mkmf.log](https://github.com/user-attachments/files/15566335/mkmf.log)

**Comment by javierm** [comment]
Hi, @jpprost :smile:.

I see the error mentions something about OpenSSL 3.3 :thinking:. This puzzles me a bit because AFAIK Ubuntu 20.04 includes OpenSSL 1.1 (you can confirm checking the installed version of the `libssl-dev` package). Could it be related to rugged being compiled with OpenSSL 3 instead of OpenSSL 1.1? :thinking:

**Reply by jpprost** [reply]
Hi,
Installed version of ```libssl-dev``` through apt: `1.1.1f-1ubuntu2.22`
But, I find ```openssl@3``` listed among the homebrew packages...
I tried as long as I could to rely on apt only, but I ended up installing homebrew because for some reason I couldn't manage to properly install and setup a ruby version manager with apt. And indeed it went quite smoothly with homebrew.

Anyway, there's a conflict here. 

Shall I simply uninstall ```openssl@3``` from brew and try to re-bundle? Hope we can rely on brew's dependencies management to prevent any side-effect.

**Reply by jpprost** [reply]
I confirm that the brew's version wass used:
`Found OpenSSL: /home/linuxbrew/.linuxbrew/Cellar/openssl@3/3.3.0/lib/libcrypto.so`
(on stdout)

**Reply by jpprost** [reply]
Hi @javierm :-)

a quick update for history sake, while I keep investigating.

The [dependency graph provided by brew](https://dreampuf.github.io/GraphvizOnline/?compressed=CYSw5gTghgDgFgAgN4CgEIERQK4BcD2AxvgHYBmGCAtAHyYC2ALBmpjgceZbZjAKYQANi3RY8RUhWp0MYQSABGhSgGIEALhD0Y8wiFysM-IdxkKBAaz6C%2BATyrAFAAQCsI3gOHTMYR-XdGnqaYfAAeMFC4AcZePBjyCqGEELYwUYZyisre8SAk2KFUcHxQwAIAzq4AdACMboYKeXggguXBGABe5bjAAY35uC1tOR0JfU2Dre2ZSqoaWjogegaiXT3tgh3Mhmu9OaEdAbvto4oBhFBUhAKDZEuRfMNxM9lqmtq6%2BhmE2c-aAWAfu0QOVhN9fjIEtplOD2toyBAAUCRt1erCcv1mlN0XFTgokRCfAlXvMPksvqIQbEZGB-oYoTBCbI6aIGUz4YjDBzpiyfH5psS5u9Fst6YpbFB6NSfMi3gtPitMAkJVKBVkhfLyYqMCRCNgIOVHmrZgg5WTRaJ8PwSOVQU4AMztC5XG4gO4XXBGwxWvg2u2OnIvDXmim8CxgK6ScDGknChWGCDmEgAN3aEGwCnsCmwLTRokTvtTgdlpJFoYwBZTMeDZe16czVGzufa7AkXATGazOcEeziyslYPzncb3d7Mh9fsEDoC9a7zZyMHDkfI0Y7DabPbTJWA8hIfBnw43Y5lsc1Fsws5H8%2BegtNpfj%2Be3u-3OV1%2BsNbQTT7yL%2BeJbjWpfqUz7VneAGigAvkAA) shows that `openssl@3` is only required by `ruby-build` through `rbenv`. 

So:
 - Option 1, I'm going to try again to install `rvm` with apt to move away from `rbenv` -- which I'll uninstall with brew, then start again the bundle and see how it goes. Hopefully it will not install openSSL3 and rely on Ubuntu's default.
 - Option 2, should option 1 not work, I will replace `rbenv` with `rvm` using homebrew, hoping that it does not depend on `openssl@3` -- which, in my opinion, is unlikely, because of the dependency to `ruby-build`.

**Reply by javierm** [reply]
I've never used brew, so I'm afraid I'm not sure. Sorry! :pray:

We've only faced the opposite problem; we had to install Ruby 3.1 (which doesn't support OpenSSL 3) on Ubuntu 22.04 (which only includes OpenSSL 3). It was... tricky :smile:. Hope this isn't as tricky!

**Comment by jpprost** [comment, marked as answer]
So, long story short: you should NOT try to combine homebrew with apt for managing your packages on Ubuntu 20.04.

I'm not ready to reproduce the error to make that statement a stronger assertion. But after trying to just replace `rbenv` with `rvm` while keeping homebrew (i.e., "Option 1" in my previous comment), I only fixed my problems when I uninstalled homebrew altogether(*). I, then, [installed the ruby version manager `rvm` with apt](https://rvm.io/rvm/install), then Ruby version `.ruby-version` with the newly-installed `rvm`.
The gems now install correctly with Bundler.

I'm now in trouble again while running `bin\rspec` (i.e., the test-suite, step #8), but I'll post a separate issue for this one if necessary.
As far as I'm concerned, the initial issue is solved, and I shall mark it so. Thank you @javierm again for the corner-stone tip on `openSSL3`.

(*) Should you need to uninstall homebrew, don't forget to remove recursively `/home/linuxbrew` and tidy up your `~/.bashrc` as well.
